src/multiselect_contextmenu.js

```diff
@@ -9,7 +9,110 @@
  */
 
 import * as Blockly from 'blockly/core';
-import {blockSelectionWeakMap, hasSelectedParent} from './global';
+import {blockSelectionWeakMap, hasSelectedParent, copyData,
+  connectionDBList, dataCopyToStorage, dataCopyFromStorage,
+  blockNumGetFromStorage} from './global';
+
+/**
+ * Copy multiple selected blocks to clipboard.
+ * @param {boolean} useCopyPasteCrossTab Whether or not to use
+ *     cross tab copy paste.
+ */
+const registerCopy = function(useCopyPasteCrossTab) {
+  const copyOptions = {
+    displayText: function(scope) {
+      let workableBlocksLength = 0;
+      const workspace = scope.block.workspace;
+      const blockSelection = blockSelectionWeakMap.get(workspace);
+      blockSelection.forEach(function(id) {
+        const block = workspace.getBlockById(id);
+        if (copyOptions.check(block)) {
+          workableBlocksLength++;
+        }
+      });
+      if (workableBlocksLength <= 1) {
+        return Blockly.Msg['CROSS_TAB_COPY']?
+        Blockly.Msg['CROSS_TAB_COPY'] : 'Copy';
+      } else {
+        return Blockly.Msg['CROSS_TAB_COPY_X_BLOCKS']?
+            Blockly.Msg['CROSS_TAB_COPY_X_BLOCKS'].replace(
+                '%1', workableBlocksLength) :
+            (Blockly.Msg['CROSS_TAB_COPY']?
+            Blockly.Msg['CROSS_TAB_COPY'] : 'Copy'
+            ) + ' (' +
+            workableBlocksLength + ')';
+      }
+    },
+    preconditionFn: function(scope) {
+      const workspace = scope.block.workspace;
+      if (workspace.options.readOnly) {
+        return 'disabled';
+      }
```

Comment:
Note that there is a discrepancy in behavior here with the paste option w.r.t. readOnly mode. Here, we return disabled, but in the paste option we return hidden. We probably also want this to be hidden (although with cross tab copy paste I think it could even be enabled since copying blocks doesn't modify the workspace).

Reply:
Thank you! I just changed this into that, for copy, we only return hidden when the workspace is read only and not in cross tab copy paste mode.